### PR TITLE
Expose environment variables from database into build commands

### LINKED ISSUE #3992: Ability to set env vars in project admin
We should allow users to inject env vars that can be maintained outside of source code. The use case here would be inputting shared secrets like API access tokens which would then be injected into the build processes. Heroku does something similar.

Eventually, we might auto generate an auth token for our api and inject this in the same way, so that Sphinx can automatically have access to our API as the user. This particular feature will likely be an API v3 feature though.

### Phase I

- [x] Add modeling for project environment variables
- [x] Add site admin UI for models
- [x] Add API response that includes environment variables so builder can consume this
- [x] Expose env variables to necessary build commands (all build commands?)

### Phase II

- [x] Add model forms for user facing interface
- [x] Add CRUD UI to project admin -- copy from established CRUD UI pattern

### PR DESCRIPTION
**Pull request by humitos**
Environment variables can be added from the Admin and they will be used when running build commands. All the variables for that particular project will be expose to all the commands.

> This is the first approach for this feature. There are more decision that can be made around this and improvements. Although, this is the easiest way to implement this since all the environment variables are exposed to all the commands (by using existing code for this) and there are no checkings to be done when running each command.

I QA this under Docker and Local building and my custom env variable was exposed in both scenarios.

### CODE REVIEW

**Comment by humitos** [comment]
If this is the path that we want to follow, I'd add tests for this new code.

**Comment by agjohnson** [comment]
This doesn't close, #3992. It only wraps up the first phase.

File: readthedocs/projects/migrations/0034_add_environment_variables.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+from django.db import migrations, models
+import django.db.models.deletion
+import django_extensions.db.fields
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('projects', '0032_increase_webhook_maxsize'),
```
Comment (agjohnson) [inline readthedocs/projects/migrations/0034_add_environment_variables.py]: what happened to 0033?

File: readthedocs/projects/models.py [hunk 2]
```diff
@@ -1109,3 +1110,19 @@ def get_feature_display(self):
         implement this behavior.
         """
         return dict(self.FEATURES).get(self.feature_id, self.feature_id)
+
+
+class EnvironmentVariable(TimeStampedModel, models.Model):
+    name = models.CharField(
+        max_length=128,
+        help_text='Name of the environment variable',
```
Comment (agjohnson) [inline readthedocs/projects/models.py]: Untranslated help text in this model

File: readthedocs/projects/tasks.py [hunk 3]
```diff
@@ -625,10 +625,15 @@ def get_env_vars(self):
...
                     self.version.slug,
-                    'bin'
+                    'bin',
                 ),
             })
 
+        # Update environment from Project's specific environment variables
+        env.update(
+            api_v2.project(self.project.pk).environment_variables().get(),
+        )
```
Comment (agjohnson) [inline readthedocs/projects/tasks.py]: Heroku limits variables from starting with "HEROKU_", should we do similar? I guess worst case the user is just breaking their own builds, there isn't anything internal we use the env vars for -- I think at least.

File: readthedocs/restapi/views/model_views.py [hunk 4]
```diff
@@ -117,6 +117,13 @@ def valid_versions(self, request, **kwargs):
             'flat': version_strings,
         })
 
+    @detail_route(permission_classes=[permissions.IsAdminUser])
+    def environment_variables(self, request, *_, **__):
+        return Response({
+            env.name: env.value
+            for env in self.get_object().environmentvariable_set.all()
+        })
```
Comment (agjohnson) [inline readthedocs/restapi/views/model_views.py]: Why make this a separate route? We already have a admin-only serializer for Project that is used for fields only used by our builders like this. I'd rather it go there and save the extra request to the API.

File: readthedocs/projects/migrations/0034_add_environment_variables.py (hunk 1 above)
Comment (humitos) [reply readthedocs/projects/migrations/0034_add_environment_variables.py]: Good catch! I had some conflics in my db/migrations and I made a mistake.

File: readthedocs/projects/tasks.py (hunk 3 above)
Comment (humitos) [reply readthedocs/projects/tasks.py]: We could.

They have more validations: https://devcenter.heroku.com/articles/config-vars#config-var-policies

The one that limits the 32kb is a good one, since if you have more data there it's very likely that you are trying to do some "weird thing".

Anyway, I think this validation should go in the Form when we allow the user to create vars from the admin tab.

File: readthedocs/restapi/views/model_views.py (hunk 4 above)
Comment (humitos) [reply readthedocs/restapi/views/model_views.py]: Good point!

File: readthedocs/projects/tasks.py (hunk 3 above)
Comment (agjohnson) [reply readthedocs/projects/tasks.py]: I'm solidly on the fence here, so feel free to take this any direction you want. We probably don't need to be too strict here.

There might be some wisdom to something like the 32kb limit (like, does docker choke on that many environment variables? I wouldn't put too much work in here, so I'm :+1: on adding anything that is easy to implement and stops users from breaking things.

File: readthedocs/projects/models.py [hunk 5]
```diff
@@ -1109,3 +1130,19 @@ def get_feature_display(self):
...
+        help_text=_('Name of the environment variable'),
+    )
+    value = models.CharField(
+        max_length=256,
+        help_text=_('Value of the environment variable'),
+    )
+    project = models.ForeignKey(
+        Project,
+        on_delete=models.CASCADE,
+        help_text=_('Project where this variable will be used'),
```
Comment (agjohnson) [inline readthedocs/projects/models.py]: Any particular thoughts on using a reverse name? Historically, we've given a verbose reverse name like "environment_variables" or something other than the default "environmentvariable_set"

File: readthedocs/restapi/serializers.py [hunk 6]
```diff
@@ -43,11 +43,6 @@ class ProjectAdminSerializer(ProjectSerializer):
         slug_field='feature_id',
     )
 
-    show_advertising = serializers.SerializerMethodField()
-
-    def get_show_advertising(self, obj):
-        return obj.show_advertising
-
```
Comment (agjohnson) [inline readthedocs/restapi/serializers.py]: I'm unfamiliar with why this change is included here, was this on purpose?

File: readthedocs/projects/tasks.py (hunk 3 above)
Comment (humitos) [reply readthedocs/projects/tasks.py]: Let do this validation on the form then, which is the easier place to implement and it will be per- env variable instead of all together.

File: readthedocs/restapi/serializers.py (hunk 6 above)
Comment (humitos) [reply readthedocs/restapi/serializers.py]: Ups, I forgot to explain why I did remove this code.

Basically, it's not necessary. That method is the default behavior of DRF when you put a field name in the `Meta.fields`.

> Alternatively names in the fields options can map to properties or methods which take no arguments that exist on the model class.

From, https://www.django-rest-framework.org/api-guide/serializers/#specifying-which-fields-to-include

File: readthedocs/projects/models.py (hunk 5 above)
Comment (humitos) [reply readthedocs/projects/models.py]: In particular in this case, I'm using `environment_variables` as a property that returns the dictionary like `{"VARIABLE": "value"}` and `environmentvariable_set.` to make queries on this relation.

File: readthedocs/projects/tasks.py (hunk 3 above)
Comment (humitos) [reply readthedocs/projects/tasks.py]: I've already started writing this form in another PR: https://github.com/rtfd/readthedocs.org/pull/4899/files#diff-49158d3dbeadb679e46430b842c441a6R786

File: readthedocs/restapi/serializers.py (hunk 6 above)
Comment (agjohnson) [reply readthedocs/restapi/serializers.py]: Roger, I'll add a QA note to our next release.

File: readthedocs/projects/models.py (hunk 5 above)
Comment (agjohnson) [reply readthedocs/projects/models.py]: This seems fair. I don't have any particular opinions or suggestions here, so probably fine to continue with a pattern like this. Perhaps there is a more Django-y pattern for verbose property name/verbose field name, instead of verbose property name/obtuse field name, but :man_shrugging:

File: readthedocs/projects/tasks.py (hunk 3 above)
Comment (agjohnson) [reply readthedocs/projects/tasks.py]: Cool, can push discussion there!

**Comment by agjohnson** [comment]
:tada:
